### PR TITLE
Improve process for importing (LGBCE) Ad-Hoc boundaries

### LINKED ISSUE #161: Improve process for importing (LGBCE) Ad-Hoc boundaries
* Moving forwards, LGBCE should be publishing shapefiles as standard once the 'final recommendations' stage is reached
* We should also maintain our own copy of these in an S3 bucket - see https://github.com/DemocracyClub/boundary-bot/issues/8
* There are 2 phases to establishing a new `DivisionSet`:
1. Create the new `Division` with names/number of posts only and `NULL` geography based on the ECO text
    * Code to do this is here: https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/organisations/management/commands/import_organisations_from_csv.py
    * Even though there can potentially be naming issues arising from typo in legislation, there's not really much we can do about it.. we just need to be able to deal with it. See #40
2. Attach boundaries to `Division` objects
    * Existing code to do this is here: https://github.com/DemocracyClub/EveryElection/blob/master/every_election/apps/organisations/management/commands/import_ad_hoc_boundaries.py
    * Need to check the LGBCE names off against the ECO names and provide some mapping between them (see below)
* At the point we create a new `DivisionSet` we should have already set an `end_date` on the previous one. As such, the logic here should be that creating a new `DivisionSet` should set the `start_date` of the new `DivisionSet` to `(end_date of the new one - 1)`. If the prevision `DivisionSet`'s `end_date` is still `NULL`, we should warn/error.
* As we move to staying on top of reviews and doing this more often, we don't want to edit `import_ad_hoc_boundaries.py` every time we need to import new files. We should be able to run commands something like like `./manage.py import_lgbce --org KHL "s3://my-bucket-name/Kingston-upon-Hull/The Kingston upon Hull (Electoral Changes) Order 2017/Hull_final_proposals.zip"` on an ad-hoc basis.

One of the things we need to deal with in order to attach boundaries from LGBCE to a `DivisionSet` created from an ECO is matching up the ward names from the ECO to the ward names in the LGBCE files. There's a few possible approaches to consider when dealing with this issue (these aren't necessarily exclusive):

* Throw an exception if they don't match and leave the user to manually edit the input file
* Compare names by levenshtein distance - make suggestions based on the closest match
* On the command line, provide a second 'mapping' file of LGBCE names --> ECO names
* We could also implement it as a file upload form in `/admin` - this might make it easier to interactively edit area names as part of the import process.

### PR DESCRIPTION
**Pull request by chris48s**
Closes #161

* I've attempted to structure this in a relatively re-usable way (e.g: I hope that if we implemented this as a file upload form in `/admin` or make more mangement command for working with Wales/Scotland data, we can re-use `DivisionSetGeographyImporter` to do the leg-work)
* If the names from the legislation don't match the names from LGBCE, we throw a `DiffException` with a `.diff` property. This allows us to output hints on the console like

```diff
- Conningbrook & Little Burton Farm
?              ^

+ Conningbrook and Little Burton Farm
?              ^^^
```

or

```diff
- Henfield
+ Henfield Ward
?         +++++
```

which makes it relatively painless to work out how to fix the issues :)

* I've uploaded all the data that has been published so far to the s3 bucket I created and tested everything we've got so far. Everything in the backlog that has been published imports cleanly (apart from Horsham - I've followed up on that with LGBCE). I'm sure as we get more data, we're going to find more edge cases we need to handle.
* There are some real-life examples of `name_map.json` at
`s3://lgbce-mirror/ashford/The%20Ashford%20%28Electoral%20Changes%29%20Order%202017/name_map.json` and
`s3://lgbce-mirror/test-valley/The%20Test%20Valley%20%28Electoral%20Changes%29%20Order%202018/name_map.json`
which are needed to get the Ashford and Test Valley data to import.
* Once we've merged and deployed #162 and imported the new division names from the legislation, we'll be able to use this to import the boundary data we have from LGBCE. the commands to do this are:

```bash
python manage.py import_lgbce ALL "allerdale/The Allerdale (Electoral Changes) Order 2017/Allerdale_final_proposals.zip"
python manage.py import_lgbce ASF "ashford/The Ashford (Electoral Changes) Order 2017/Ashford_final_proposals.zip" -n "Ward_name"
python manage.py import_lgbce CRY "croydon/London Borough of Croydon (Electoral Changes) Order 2017/Croydon_final_proposals.zip"
python manage.py import_lgbce HAO "harborough/The Harborough (Electoral Changes) Order 2017/Harborough_polygons.zip"
python manage.py import_lgbce KHL "kingston-upon-hull/The Kingston upon Hull (Electoral Changes) Order 2017/Hull_final_proposals.zip"
python manage.py import_lgbce NEC "newcastle-under-lyme/Newcastle-under-Lyme (Electoral Changes) Order 2017/NEWCASTLEUNDERLYME_final_proposals.shp.zip" -n "Ward_name"
python manage.py import_lgbce NNO "north-norfolk/The North Norfolk (Electoral Changes) Order 2017/NorthNorfolk_final_proposals.zip"
python manage.py import_lgbce TEN "tendring/The Tendring (Electoral Changes) Order 2017/Tendring_final_proposals.zip"
python manage.py import_lgbce TES "test-valley/The Test Valley (Electoral Changes) Order 2018/TestValley_final_proposals.zip"
python manage.py import_lgbce TEW "tewkesbury/The Tewkesbury (Electoral Changes) Order 2018/Tewkesbury_final_proposals.zip"
```

Horsham is waiting on a correction. Everywhere else hasn't been published yet.

* I haven't written any tests for any of this (yet) because I am a bad person :(

### CODE REVIEW

**Comment by chris48s** [comment]
Having received correction from LGBCE, Horsham will be

```bash
python manage.py import_lgbce HOR "horsham/Horsham (Electoral Changes) Order 2017/Horsham-fixed.zip"
```

**Comment by chris48s** [comment]
Note to self: outstanding query with LGBCE on Ashford - don't run that one if we merge this before hearing back from them

File: every_election/apps/organisations/management/commands/import_lgbce.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                raise
+
+    def get_division_set(self, org_code):
+        org = Organisation.objects.get(official_identifier=org_code)
+        divsets = OrganisationDivisionSet.objects\
+            .filter(organisation=org)\
+            .order_by('-start_date')
+
+        # divsets[0] is the DivisionSet with the most recent start date
+        if not divsets[0].divisions.all():
```
Comment (symroe) [inline every_election/apps/organisations/management/commands/import_lgbce.py]: I think we should define `get_latest_by` on the model and use `latest` here, just in case we end up changing the default ordering for some other reason in the future. I understand this might break for other reasons anyway (it will have an end date so line 107 will fail), but that might make the error harder to debug if we don't know why we don't have the latest model.

File: every_election/apps/organisations/importers.py [hunk 2]
```diff
@@ -1,12 +1,101 @@
...
+        super().__init__(message)
+        self.diff = diff
+
+
+class DivisionSetGeographyImporter:
+
+    def __init__(self, data, division_set, name_column='name', name_map={}, srid=27700):
+        if not isinstance(data, DataSource):
+            error = "param 'data' must be an instance of django.contrib.gis.gdal.DataSource"
+            raise TypeError(error)
```
[the remark this answers is not in the document]
Comment (chris48s) [reply every_election/apps/organisations/importers.py]: eeeeeeeeh.. I was optomistic about this, but its not as useful as you think. The operative word is very much **hint**. For example, if I define

```python
def somefunction(param: int):
    return "foo"
```

`somefunction("blatantly a string")` returns "foo" rather than throwing an exception. You have to bolt something on to actually  **enforce** checking. There will probably be a module at some point the does some wizardry with the AST so you can `pip install actually-enforce-type-hints` and it will do it transparently without having to remember to wrap every single function where you've used a type hint in a decorator for it to actually do anything. I think the base implementation in 3.5+ is too soft to be helpful.

Could be worse though. The javascript solution to adding static type hints to a dynamic language is to ignore it in the core language and then 2 huge companies go and develop competing 2 frameworks to bolt it on in userland, both with pretty much identical design goals but which are completely incompatible with each other - because javascript :D

Not often I'll say this, but for "how to add static type hints to a dynamic language and not make a complete pigs ear of it", see PHP 7.1. They nailed it.

Bit of a tangent there, but when you say "looking forward", are you still using 3.4 locally? I think once we get EE and website off DC-Prod everything will be running 3.5 in prod.

File: every_election/apps/organisations/management/commands/import_lgbce.py (hunk 1 above)
Comment (chris48s) [reply every_election/apps/organisations/management/commands/import_lgbce.py]: Yeah good plan. I've used this same approach in both this PR and #162 so what I've done is pushed 94b2b69 and b6ebf7c on to that branch because #162 is more important and already has tests.

If you prioritise the review of #162, I can rebase this PR and #145 (which will now have a migration conflict because both #162 and #145 now define a migration `0029` on the same model) onto it once that is merged.

Comment (chris48s) [reply every_election/apps/organisations/management/commands/import_lgbce.py]: In the meantime, I can crack on with adding tests for this as it is  - then it will be easier to ensure I'm not breaking anything when I rebase it on to that anyway

File: every_election/apps/organisations/importers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (symroe) [reply every_election/apps/organisations/importers.py]: Good info! I actually only saw the PEP though http://mypy-lang.org/, maybe that addresses some of your comments?

*Anyway* 

Yes, also, I'm too focused on YNR's 2.7ness that I'm forgetting we're on >=3.5 in most other places

**Comment by chris48s** [comment]
...aaand the rest of the backlog will be covered by:

```bash
python manage.py import_lgbce BBD "blackburn-and-darwen/The Blackburn with Darwen (Electoral Changes) Order 2017/Blackburn with Darwen.zip"
python manage.py import_lgbce BOS "bolsover/The Bolsover (Electoral Changes) Order 2017/Bolsover.zip" -n "wardname"
python manage.py import_lgbce NED "north-east-derbyshire/The North East Derbyshire (Electoral Changes) Order 2017/North East Derbyshire.zip" -n "wardname"
python manage.py import_lgbce RIB "ribble-valley/The Ribble Valley (Electoral Changes) Order 2017/Ribble Valley.zip" -n "ward_name"
python manage.py import_lgbce SLA "south-lakeland/The South Lakeland (Electoral Changes) Order 2017/South Lakeland.zip"
python manage.py import_lgbce HUN "huntingdonshire/The Huntingdonshire (Electoral Changes) Order 2017/Huntingdonshire.zip"
python manage.py import_lgbce SNO "south-norfolk/The South Norfolk (Electoral Changes) Order 2017/South Norfolk.zip"
python manage.py import_lgbce CHI "chichester/The Chichester (Electoral Changes) Order 2017/Chichester.zip"
python manage.py import_lgbce HAG "harrogate/The Harrogate (Electoral Changes) Order 2017/Harrogate.zip"
python manage.py import_lgbce LDS "leeds/The Leeds (Electoral Changes) Order 2017/Leeds.zip" -n "wardname"
python manage.py import_lgbce NET "newcastle-upon-tyne/The Newcastle upon Tyne (Electoral Changes) Order 2017/Newcastle upon Tyne.zip"
python manage.py import_lgbce SUR "surrey-heath/The Surrey Heath (Electoral Changes) Order 2017/Surrey Heath_fixed.zip"
python manage.py import_lgbce TOR "torridge/The Torridge (Electoral Changes) Order 2017/Torridge_fixed.zip"
python manage.py import_lgbce BEX "bexley/London Borough of Bexley (Electoral Changes) Order 2017/Bexley_fixed.zip"
python manage.py import_lgbce RDB "redbridge/The London Borough of Redbridge (Electoral Changes) Order 2017/Redbridge_fixed.zip"
python manage.py import_lgbce SCA "south-cambridgeshire/South Cambridgeshire (Electoral Changes) Order 2017/South Cambridgeshire_fixed.zip"
python manage.py import_lgbce TEI "teignbridge/The Teignbridge (Electoral Changes) Order 2017/Teignbridge.zip"
```
